### DISCUSSION TITLE
Is there any way to set warmstart externally?

### DISCUSSION THREAD

**Discussion by Sevarius** [coin-or/Sonnet#10] [Q&A] [answered]
The question is simple, but the answer is probably not.
Is there a way to set warmstart if I have array of variable values with valid solution?

I saw, there is a way to set warmstart using another solved problem
```
/// The WarmStart object should be taken from the solver (GetWarmStart) just after a Solve(),
/// and applied to the solver (SetWarmStart) before another solve.
/// Use via solver.GetWarmStart() and solver.SetWarmStart().
```
but if I don't have one and solving this problem without warmstart is far too long...

**Comment by jhmgoossens** [comment, marked as answer]
Are you solving an LP or a MIP? 

WarmStart is about kick-starting an LP -- but that isn't very obvious from the docs.
If you're talking about a feasible solution for a MIP, then I believe the way is either via 
1) Cbc's "mipstart" via a file
For this if you can make the solution file, you can tell Sonnet via CbcSolverArgs to tell Cbc to use mipstart. For an example of how to add solver args, see [SonnetTest41](https://github.com/coin-or/SONNET/blob/13e872476cbb9e798832d06fbe06f8c01bbcd68b/test/SonnetTest/Sonnet_BasicTests.cs#L2687). See also the mipstart option of the cbc command line tool (mipstart?)  

or 

2) straight via  CbcModel's setBestSolution and setCutoff etc. 
For this, SonnetWrapper doesn't include the necessary code (yet). For an example, see [CbcSolver](https://github.com/coin-or/Cbc/blob/879602724a65987c5cfb0b9fbacfa192c93df42c/src/CbcSolver.cpp#L4143) 

Does that help?

**Reply by Sevarius** [reply]
Sorry for not mentioning this in the question, I am solving MIP problem.

Thanks you very much for your answer. 
I'm looking forward to the new version of this library.
Until that, I'll probably try to use your second solution and will DLLImport some C++ functions (if I'm not mistaken)

**Reply by jhmgoossens** [reply]
If you want to try 2. then IMO simply extending SonnetWrapper for [CbcModel](https://github.com/coin-or/SONNET/blob/master/src/SonnetWrapper/CbcModel.h)  is a lot easier than trying with DLLImport. 
Let me know how it goes and which methods you end up needing?

**Reply by Sevarius** [reply]
Wow, I just realized what you offered me...
I need some time to compile all the steps I shoud take to do this, but your solution is unconditionally better.

This will be my first experience changing and bilding such a complex project)

**Reply by jhmgoossens** [reply]
It’s really not that difficult :-)
You don’t need more than VS2019 and the source code of Sonnet and the related COIN projects. 

Start with reading the Sonnet [Install](https://github.com/coin-or/SONNET/blob/master/INSTALL.txt) instructions. (Parts for cbc sources look out of date.)

I realise now how much easier building this would be if I’d include all the sources in the Sonnet binary distribution. I’ll look into that.

**Reply by Sevarius** [reply]
Yeah, the function code is not that complex (I looked one from [here](https://github.com/coin-or/Cbc/blob/0d460fd66f36253cbfdb15c5d81636dabbeb94ce/Cbc/src/Cbc_C_Interface.cpp#L625)), but I ran into that there is no source code for VS2019 in your [link](http://www.coin-or.org/download/source/Cbc/), even in the latest version.

I've tried to change reference for VS2017 (v14 folder), but it turned out, that there were no some .vcxproj files, like libOsiCbc.vcxproj.

Is there any way to get VS2019 source files?

**Reply by jhmgoossens** [reply]
I just made this a lot easier by adding the full src (sonnet and COIN-OR libs) as AppVeyor artefact, and update the install.txt.

Getting the source code
---------------------
Option 1: Get the full src.zip artifact from AppVeyor (if available)
The easiest way to get started with SONNET source code is to use the ..-src.zip artefact (if available) from the latest AppVeyor build
https://ci.appveyor.com/project/coin-or/sonnet/build/artifacts
This zip contains the SONNET sources as well as all COIN-OR dependencies from their master branch. 
Simply unzip, open the SONNET solution in sonnet\MSVisualStudio\v16\Sonnet.sln, and Build.

**Reply by Sevarius** [reply]
Wow, Thanks you very much!

But it seems like there are some errors in the CoinMessageHandler.h

```
line 98: Error E1971 class "CoinMessageHandler" (declared at line 344 of "Sonnet-master-CI-src/CoinUtils/src/CoinMessageHandler.hpp") was previously declared as a different kind of class

line 182: Error E0265 function "COIN::CoinMessageHandlerProxy::printBase" (declared at line 136 of "SONNET-MASTER-CI-SRC\SONNET\SRC\SONNETWRAPPER\CoinMessageHandler.h") is inaccessible
```
Second one I can fix, by simply making function public, but the first one I really don't know how to fix.

**Reply by Sevarius** [reply]
But it's building anyway!

Thank you very much for your help)

**Reply by jhmgoossens** [reply]
I just realize SonnetWrapper already wraps CbcModel's setObjValue to provide an objective function value before solving. See [Sonnet_BasicTests for an example](https://github.com/coin-or/SONNET/blob/7afddf003f2f5a0fb5ec6ef58452dd2b22d45e0b/test/SonnetTest/Sonnet_BasicTests.cs#L2804).

**Comment by Sevarius** [comment]
I wrote two methods, which, i think, will be very useful in Sonnet:

```
/// <summary>
/// Set initial solution
/// </summary>
/// <param name="solution">array of initial solution coefficients</param>
void setWarmStart(array<double>^ solution)
{
	int n = Base->getNumCols();
	const double* objvec = Base->solver()->getObjCoefficients();
	double objval = 0;
	for (int i = 0; i < n; i++) {
		objval += objvec[i] * solution[i];
	}
	pin_ptr<double> pointer = &solution[0];
	Base->setBestSolution(pointer, n, objval, true);
}

/// <summary>
/// Set maximum seconds
/// </summary>
/// <param name="maxSeconds">max seconds to solve</param>
void setMaxSeconds(double maxSeconds)
{
	Base->setDblParam(Base->CbcMaximumSeconds, maxSeconds);
}
```
But I have a suspicion, that setBestSolution did not force solver to start searching from the solution, that I passed to it.

I made a problem that solves in 30 seconds, I thought that, if I pass optimal solution to it, solver will start his searching point from this solution and will solve faster, but it didn't do that.

@jhmgoossens could you please tell me if I understand the principle of WarmStart in Cbc correctly? Or it is mathematically impossible to set the starting point of the algorithm?

**Reply by jhmgoossens** [reply]
When you provide a feasible solution to the MIP solver before solving--similar to how several heuristics are used at start, or actually similar to any integer feasible solution found throughout the search process--then the objective function value is used during the search tree to cut off (prune) parts of the search tree where the relaxation solution is already worse than the objective function value of the best known solution. For those parts of the search tree, you'd be sure no better solution can be found. That's 'all'. 
If you're solving a problem for which the (LP) relaxation is not tight/large gap, then having a feasible integer solution doesn't actually help all that much and still a large part of the tree needs to be searched since each part of the tree will seem very promising since the LP relaxation is not tight. 
So your millage may vary. Does that make sense?

**Reply by Sevarius** [reply]
Yes, thank you very much for the answer! 
Probably, I have bad example, hope that in real problems setBestSolution will works better.
